File: packages/ai/src/codebase/classifier.ts
/**
 * File classifier for codebase ingestion.
 *
 * Classifies files by:
 * - language (from extension)
 * - domain (from path segments)
 * - artifactType (API/model/config/test/doc/source)
 *
 * Also detects secret files that should be skipped.
 */

/**
 * Classification result for a single file.
 */
export interface FileClassification {
  domain: string
  artifactType: string
  language: string
}

const LANG_MAP: Record<string, string> = {
  ts: 'typescript',
  tsx: 'typescript',
  js: 'javascript',
  jsx: 'javascript',
  py: 'python',
  go: 'go',
  md: 'markdown',
  yaml: 'yaml',
  yml: 'yaml',
  json: 'json',
}

const DOMAIN_SEGMENTS = new Set([
  'api',
  'models',
  'services',
  'utils',
  'components',
  'hooks',
  'lib',
  'routes',
  'controllers',
  'middleware',
  'helpers',
  'types',
  'schema',
  'db',
  'database',
])

/**
 * Classify a file by path.
 *
 * @param filePath - Relative or absolute file path
 * @returns Classification: domain, artifactType, language
 */
export function classifyFile(filePath: string): FileClassification {
  const parts = filePath.replace(/\\/g, '/').split('/')
  const ext = filePath.split('.').pop()?.toLowerCase() ?? ''

  // Language from extension
  const language = LANG_MAP[ext] ?? 'unknown'

  // Domain: first matching path segment
  const domain =
    parts.find((p) => DOMAIN_SEGMENTS.has(p.toLowerCase())) ?? 'general'

  // Artifact type from path and extension
  let artifactType = 'source'
  const lower = filePath.toLowerCase()

  if (lower.includes('.test.') || lower.includes('.spec.') || lower.includes('/test') || lower.includes('/tests') || lower.includes('/__tests__')) {
    artifactType = 'test'
  } else if (lower.includes('config') || ext === 'yaml' || ext === 'yml' || ext === 'json') {
    artifactType = 'config'
  } else if (ext === 'md') {
    artifactType = 'doc'
  } else if (lower.includes('model') || lower.includes('schema')) {
    artifactType = 'model'
  } else if (lower.includes('api') || lower.includes('route') || lower.includes('controller') || lower.includes('endpoint')) {
    artifactType = 'api'
  }

  return { domain, artifactType, language }
}

/**
 * Detect whether a file should be skipped because it likely contains secrets.
 *
 * Skips if:
 * - Path contains `.env`
 * - Content matches API_KEY=, password:, or SECRET= patterns
 *
 * @param filePath - File path (relative or absolute)
 * @param content  - File content as a string
 * @returns true if the file should be excluded
 */
export function isSecretFile(filePath: string, content: string): boolean {
  if (filePath.includes('.env')) return true
  if (/API_KEY\s*=|password\s*:|SECRET\s*=/i.test(content)) return true
  return false
}

/**
 * The allowed source-code and text extensions for indexing.
 */
export const ALLOWED_EXTENSIONS = new Set([
  'ts', 'tsx', 'js', 'jsx', 'py', 'go', 'md', 'yaml', 'yml', 'json',
])

File: packages/gamole_ai/codebase/classifier.py
"""File classifier - ported from packages/ai/src/codebase/classifier.ts."""

import re
from dataclasses import dataclass

LANG_MAP: dict[str, str] = {
    "ts": "typescript",
    "tsx": "tsx",
    "js": "javascript",
    "jsx": "javascript",
    "py": "python",
    "go": "go",
    "md": "markdown",
    "yaml": "yaml",
    "yml": "yaml",
    "json": "json",
}

DOMAIN_SEGMENTS = {
    "api", "models", "services", "utils", "components", "hooks", "lib",
    "routes", "controllers", "middleware", "helpers", "types", "schema",
    "db", "database",
}

ALLOWED_EXTENSIONS = {
    "ts", "tsx", "js", "jsx", "py", "go", "md", "yaml", "yml", "json",
}


def detect_language(file_path: str) -> str | None:
    """Return tree-sitter language name for a file path, or None if unsupported.

    Maps file extensions to tree-sitter language names.
    Note: .tsx maps to 'tsx' (separate parser from 'typescript').
    """
    ext = file_path.rsplit(".", 1)[-1].lower() if "." in file_path else ""
    return LANG_MAP.get(ext)

@dataclass
class FileClassification:
    domain: str
    artifact_type: str
    language: str


def classifyFile(file_path: str) -> FileClassification:
    """Classify a file by path."""
    parts = file_path.replace("\\", "/").split("/")
    ext = file_path.rsplit(".", 1)[-1].lower() if "." in file_path else ""

    language = LANG_MAP.get(ext, "unknown")

    domain = next((p for p in parts if p.lower() in DOMAIN_SEGMENTS), "general")

    artifact_type = "source"
    lower = file_path.lower()

    if any(x in lower for x in [".test.", ".spec.", "/test", "/tests", "/__tests__"]):
        artifact_type = "test"
    elif "config" in lower or ext in ("yaml", "yml", "json"):
        artifact_type = "config"
    elif ext == "md":
        artifact_type = "doc"
    elif "model" in lower or "schema" in lower:
        artifact_type = "model"
    elif any(x in lower for x in ["api", "route", "controller", "endpoint"]):
        artifact_type = "api"

    return FileClassification(domain=domain, artifact_type=artifact_type, language=language)


def isSecretFile(file_path: str, content: str) -> bool:
    """Detect whether a file should be skipped because it likely contains secrets."""
    if ".env" in file_path:
        return True
    if re.search(r"API_KEY\s*=|password\s*:|SECRET\s*=", content, re.IGNORECASE):
        return True
    return False
